rust_icu_umsg/src/lib.rs

```diff
@@ -0,0 +1,211 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// The message format C api is variadic.
+#![feature(c_variadic)]
+
+//! # Implementation of the functions in the ICU4C `umsg.h` header.
+//!
+//! This is the support for `MessageFormat` based international message formatting.
+//!
+
+use {
+    anyhow::anyhow, rust_icu_common as common, rust_icu_sys as sys, rust_icu_sys::*,
+    rust_icu_uloc as uloc, rust_icu_ustring as ustring, std::convert::TryFrom,
+};
+
+/// A zero-value parse error, used to initialize types that get passed into FFI code.
+static NO_PARSE_ERROR: sys::UParseError = sys::UParseError {
+    line: 0,
+    offset: 0,
+    preContext: [0; 16usize],
+    postContext: [0; 16usize],
+};
+
+/// Converts a parse error to a Result.
+fn parse_ok(e: sys::UParseError) -> Result<(), common::Error> {
+    if e == NO_PARSE_ERROR {
+        return Ok(());
+    }
+    Err(common::Error::Wrapper(anyhow!(
+        "parse error: line: {}, offset: {}",
+        e.line,
+        e.offset
+    )))
+}
+
+/// The implementation of the ICU `UMessageFormat*`.
+///
+/// Implements `UMessageFormat`
+#[derive(Debug)]
+pub struct UMessageFormat {
+    rep: *mut sys::UMessageFormat,
+}
+
+impl Drop for UMessageFormat {
+    /// Implements `umsg_close`.
+    fn drop(&mut self) {
+        unsafe {
+            versioned_function!(umsg_close)(self.rep);
+        }
+    }
+}
+
+/// The formatting arguments to be used in the call to `UMessageFormat::format`.
+///
+/// According to the [ICU documentation for `umsg_format`](https://unicode-org.github.io/icu-docs/apidoc/released/icu4c/umsg_8h.html#a90a4b5fe778754e5da52f7c2e5fd6048):
+///
+/// > for all numeric arguments double is assumed unless the type is explicitly
+/// > integer (long).  All choice format arguments must be of type double.
+///
+/// Additionally, we determined by inspection that the string format is necessarily
+/// `rust_icu_ustring::UChar`.
+pub enum FmtArg {
```

Comment:
Please implement `From<f64>`, `From<i32>`, etc., for `FmtArg`. For the string variants, please implement conversions from `&str` and `String`.

Reply:
I rethought the formatting, and looks like `FmtArg` is not needed.